crates/uv/tests/pip_compile.rs

```diff
@@ -12382,3 +12382,59 @@ fn prune_unreachable() -> Result<()> {
 
     Ok(())
 }
+
+/// Resolve a package from an optional extra in a `pyproject.toml` file.
+#[test]
+fn unsafe_index_error() -> Result<()> {
+    let context = TestContext::new("3.12");
+    let pyproject_toml = context.temp_dir.child("pyproject.toml");
+    pyproject_toml.write_str(
+        r#"[project]
+name = "project"
+requires-python = ">=3.12"
+dependencies = ["requests==2.32.3"]
+
+[tool.uv]
+extra-index-url = ["https://download.pytorch.org/whl/cu121"]
+"#,
+    )?;
+
+    uv_snapshot!(context.filters(), context.pip_compile()
+            .arg("pyproject.toml"), @r###"
+    success: false
+    exit_code: 1
+    ----- stdout -----
+
+    ----- stderr -----
+      × No solution found when resolving dependencies:
+      ╰─▶ Because there is no version of requests==2.32.3 and project depends on requests==2.32.3, we can conclude that your requirements are unsatisfiable.
+
+          hint: `requests` was found on https://download.pytorch.org/whl/cu121, but not at the requested version (requests==2.32.3). A compatible version may be available on a subsequent index (e.g., https://pypi.org/simple). If both indexes are equally trusted, use `--index-strategy unsafe-best-match` to consider all indexes that list `requests`.
```

Comment:
I'm not sure if we want to include more info here on what our default behavior is and why. I think "If both indexes are equally trusted" and the name `unsafe-best-match` imply that it's for security reasons, so I left it as-is.

Reply:
I think we should provide a bit more information here. Verbose is better than not in these hints.

We never link to the documentation from the CLI right now, but we might want to do that in the future.